packages/powersync/lib/src/streaming_sync.dart

```diff
@@ -159,40 +164,49 @@ class StreamingSyncImplementation {
   }
 
   Future<void> uploadAllCrud() async {
+    // Keep track of the first item in the CRUD queue for the last `uploadCrud` iteration.
+    CrudEntry? checkedCrudItem;
+
     while (true) {
       try {
-        bool done = await uploadCrudBatch();
-        _updateStatus(uploadError: _noError);
-        if (done) {
+        // This is the first item in the FIFO CRUD queue.
+        CrudEntry? nextCrudItem = await adapter.nextCrudItem();
+        if (nextCrudItem != null) {
+          if (nextCrudItem.clientId == checkedCrudItem?.clientId) {
+            // This will force a higher log level than exceptions which are caught here.
+            isolateLogger.warning(
+                """Potentially previously uploaded CRUD entries are still present in the upload queue. 
+                Make sure to handle uploads and complete CRUD transactions or batches by calling and awaiting their [.complete()] method.
+                The next upload iteration will be delayed.""");
+            throw Exception(
+                'Delaying due to previously encountered CRUD item.');
+          }
+
+          checkedCrudItem = nextCrudItem;
+          await uploadCrud();
+          _updateStatus(uploadError: _noError);
+        } else {
+          // Uploading is completed
+          await adapter.updateLocalTarget(() => getWriteCheckpoint());
           break;
         }
       } catch (e, stacktrace) {
+        checkedCrudItem = null;
         isolateLogger.warning('Data upload error', e, stacktrace);
         _updateStatus(uploading: false, uploadError: e);
         await Future.delayed(retryDelay);
-      }
-    }
-    _updateStatus(uploading: false);
-  }
-
-  Future<bool> uploadCrudBatch() async {
-    return crudMutex.lock(() async {
-      if ((await adapter.hasCrud())) {
-        _updateStatus(uploading: true);
```

Comment:
We should not remove the `crudMutex` locking here. We should also keep the status update for `uploading: true`